File: app/src/main/java/com/frolo/muse/engine/AudioFx.java
package com.frolo.muse.engine;

import com.frolo.muse.model.preset.NativePreset;
import com.frolo.muse.model.preset.Preset;
import com.frolo.muse.model.reverb.Reverb;

import org.jetbrains.annotations.Nullable;

import java.util.List;


public interface AudioFx {
    /**
     * Save the current state of audio fx in own storage.
     */
    @Deprecated()
    void save();

    /**
     * Registers a new [observer].
     * This method has no effect if there is same observer registered.
     */
    void registerObserver(AudioFxObserver observer);
    /**
     * Unregisters [observer].
     * This method has no effect if the observer isn't registered.
     */
    void unregisterObserver(AudioFxObserver observer);

    /**
     * Returns true if the audio fx is initialized and has correct state, false - otherwise.
     */
    boolean isAvailable();
    /**
     * Returns true if the audio fx is enabled, false - otherwise.
     */
    boolean isEnabled();
    /**
     * Enables the audio fx if [enabled] is true, disables - otherwise.
     */
    void setEnabled(boolean enabled);

    /**
     * Returns true if the audio fx has equalizer, false - otherwise.
     */
    boolean hasEqualizer();
    /**
     * Returns the lower border of equalizer band range.
     * It will return 0 if the audio fx has no equalizer.
     */
    short getMinBandLevelRange();
    /**
     * Returns the higher border of equalizer band range.
     * It will return 0 if the audio fx has no equalizer.
     */
    short getMaxBandLevelRange();
    /**
     * Returns an array of frequency ranges for [band] in equalizer.
     * The size of array must be 2.
     * Implementation at discretion if the audio fx has no equalizer.
     */
    int[] getBandFreqRange(short band);
    /**
     * Returns number of bands in equalizer.
     * It will return 0 if the audio fx has no equalizer.
     */
    short getNumberOfBands();
    /**
     * Returns the level value of [band].
     * It will return 0 if the audio fx has no equalizer.
     */
    short getBandLevel(short band);
    /**
     * Sets the level value of [band] to [level].
     * It will have no effect if the audio fx has no equalizer.
     */
    void setBandLevel(short band, short level);

    /**
     * Returns list of available native presets.
     * @return list of available native presets.
     */
    List<NativePreset> getNativePresets();

    /**
     * Returns the preset that is being currently used in this AudioFx.
     * @return the current preset, or null if no preset is being used
     */
    @Nullable
    Preset getCurrentPreset();
    /**
     * Makes audio fx use the given <code>preset</code>.
     * @param preset to use
     */
    void usePreset(Preset preset);
    /**
     * Forces audio fx unuse preset.
     * It will have no effect if the audio fx has no equalizer.
     */
    void unusePreset();

    /**
     * Returns true if the audio fx has preset reverb effect, false - otherwise.
     */
    boolean hasPresetReverbEffect();

    /**
     * Returns list of available reverbs.
     * @return list of available reverbs.
     */
    List<Reverb> getReverbs();
    /**
     * Returns the reverb effect that is being currently used in this AudioFx.
     * @return the current reverb effect
     */
    Reverb getCurrentReverb();
    /**
     * Make this AudioFx use the given <code>reverb</code> effect.
     * @param reverb to use
     */
    void useReverb(Reverb reverb);

    /**
     * Returns true if the audio fx has bass boost effect, false - otherwise.
     */
    boolean hasBassBoost();
    /**
     * Returns the lower border of bass boost strength.
     * It will return 0 if the audio fx has no bass boost effect.
     */
    short getMinBassStrength();
    /**
     * Returns the higher border of bass boost strength.
     * It will return 0 if the audio fx has no bass boost effect.
     */
    short getMaxBassStrength();
    /**
     * Returns the current strength of bass boost effect.
     * It will return 0 if the audio fx has no bass boost effect.
     */
    short getBassStrength();
    /**
     * Sets the strength value of bass boost effect to [strength].
     * It will return 0 if the audio fx has no bass boost effect.
     */
    void setBassStrength(short strength);

    /**
     * Returns true if the audio fx has virtualizer effect, false - otherwise.
     */
    boolean hasVirtualizer();
    /**
     * Returns the lower border of virtualizer strength.
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    short getMinVirtualizerStrength();
    /**
     * Returns the higher border of virtualizer strength.
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    short getMaxVirtualizerStrength();
    /**
     * Returns the current strength of virtualizer strength.
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    short getVirtualizerStrength();
    /**
     * Sets the strength value of virtualizer effect to [strength].
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    void setVirtualizerStrength(short strength);
}

File: app/src/main/java/com/frolo/muse/engine/AudioFx.kt
package com.frolo.muse.engine

import com.frolo.muse.model.preset.CustomPreset
import com.frolo.muse.model.preset.NativePreset


interface AudioFx {
    /**
     * Save the current state of audio fx in own storage.
     */
    @Deprecated("Consider saving state after each mutation")
    fun save()

    /**
     * Registers a new [observer].
     * This method has no effect if there is same observer registered.
     */
    fun registerObserver(observer: AudioFxObserver)
    /**
     * Unregisters [observer].
     * This method has no effect if the observer isn't registered.
     */
    fun unregisterObserver(observer: AudioFxObserver)

    /**
     * Returns true if the audio fx is initialized and has correct state, false - otherwise.
     */
    fun isAvailable(): Boolean
    /**
     * Returns true if the audio fx is enabled, false - otherwise.
     */
    fun isEnabled(): Boolean
    /**
     * Enables the audio fx if [enabled] is true, disables - otherwise.
     */
    fun setEnabled(enabled: Boolean)

    /**
     * Returns true if the audio fx has equalizer, false - otherwise.
     */
    fun hasEqualizer(): Boolean
    /**
     * Returns the lower border of equalizer band range.
     * It will return 0 if the audio fx has no equalizer.
     */
    fun getMinBandLevelRange(): Short
    /**
     * Returns the higher border of equalizer band range.
     * It will return 0 if the audio fx has no equalizer.
     */
    fun getMaxBandLevelRange(): Short
    /**
     * Returns an array of frequency ranges for [band] in equalizer.
     * The size of array must be 2.
     * Implementation at discretion if the audio fx has no equalizer.
     */
    fun getBandFreqRange(band: Short): IntArray
    /**
     * Returns number of bands in equalizer.
     * It will return 0 if the audio fx has no equalizer.
     */
    fun getNumberOfBands(): Short
    /**
     * Returns the level value of [band].
     * It will return 0 if the audio fx has no equalizer.
     */
    fun getBandLevel(band: Short): Short
    /**
     * Sets the level value of [band] to [level].
     * It will have no effect if the audio fx has no equalizer.
     */
    fun setBandLevel(band: Short, level: Short)

    /**
     * Returns number of presets in equalizer.
     * It will return 0 if the audio fx has no equalizer.
     */
    fun getNumberOfPresets(): Short
    /**
     * Returns the name of band at [index] position in equalizer.
     * It will return empty string if the audio fx has no equalizer.
     */
    fun getPresetName(index: Short): String
    /**
     * Forces audio fx use native [preset].
     * It will have no effect if the audio fx has no equalizer.
     */
    fun useNativePreset(preset: NativePreset)
    /**
     * Forces audio fx unuse preset.
     * It will have no effect if the audio fx has no equalizer.
     */
    fun unusePreset()
    /**
     * Forces audio fx use custom [preset].
     * It will have no effect if the audio fx has no equalizer.
     */
    fun useCustomPreset(preset: CustomPreset)
    /**
     * Returns true if audio fx equalizer is using native preset.
     * It will return false if the audio fx is currently not using native preset or has no equalizer at all.
     */
    fun isUsingNativePreset(): Boolean
    /**
     * Returns true if audio fx equalizer is custom native preset.
     * It will return false if the audio fx is currently not using custom preset or has no equalizer at all.
     */
    fun isUsingCustomPreset(): Boolean
    /**
     * Returns native preset that the audio fx equalizer is currently using.
     * It will return null if the audio is currently not using any native preset or has no equalizer at all.
     */
    fun getCurrentNativePreset(): NativePreset?
    /**
     * Returns custom preset that the audio fx equalizer is currently using.
     * It will return null if the audio is currently not using any custom preset or has no equalizer at all.
     */
    fun getCurrentCustomPreset(): CustomPreset?
    /**
     * Returns true if the audio fx has preset reverb effect, false - otherwise.
     */
    fun hasPresetReverb(): Boolean
    /**
     * Returns number of available preset reverbs.
     * It will return 0 if the audio fx has no preset reverb effect.
     */
    fun getNumberOfPresetReverbs(): Short
    /**
     * Returns indexes of available preset reverbs.
     * It will return an empty array if the audio fx has no preset reverb effect.
     */
    fun getPresetReverbIndexes(): ShortArray
    /**
     * Returns the name of preset reverb at [index] position.
     * It will return empty string if the audio fx has no preset reverb effect.
     */
    fun getPresetReverbName(index: Short): String
    /**
     * Force the audio fx preset reverb effect use the preset at [index] position.
     * It will have no effect if the audio fx has no preset reverb effect.
     */
    fun usePresetReverb(index: Short)
    /**
     * Returns index of preset reverb that the audio fx preset reverb effect is currently using.
     * It will return -1 if the audio fx has no preset reverb effect.
     */
    fun getCurrentPresetReverb(): Short

    /**
     * Returns true if the audio fx has bass boost effect, false - otherwise.
     */
    fun hasBassBoost(): Boolean
    /**
     * Returns the lower border of bass boost strength.
     * It will return 0 if the audio fx has no bass boost effect.
     */
    fun getMinBassStrength(): Short
    /**
     * Returns the higher border of bass boost strength.
     * It will return 0 if the audio fx has no bass boost effect.
     */
    fun getMaxBassStrength(): Short
    /**
     * Returns the current strength of bass boost effect.
     * It will return 0 if the audio fx has no bass boost effect.
     */
    fun getBassStrength(): Short
    /**
     * Sets the strength value of bass boost effect to [strength].
     * It will return 0 if the audio fx has no bass boost effect.
     */
    fun setBassStrength(strength: Short)

    /**
     * Returns true if the audio fx has virtualizer effect, false - otherwise.
     */
    fun hasVirtualizer(): Boolean
    /**
     * Returns the lower border of virtualizer strength.
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    fun getMinVirtualizerStrength(): Short
    /**
     * Returns the higher border of virtualizer strength.
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    fun getMaxVirtualizerStrength(): Short
    /**
     * Returns the current strength of virtualizer strength.
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    fun getVirtualizerStrength(): Short
    /**
     * Sets the strength value of virtualizer effect to [strength].
     * It will return 0 if the audio fx has no virtualizer effect.
     */
    fun setVirtualizerStrength(strength: Short)
}